vpr/src/route/rr_graph_fpga_interchange.cpp

```diff
@@ -0,0 +1,1259 @@
+        : ar_(arch_reader)
+        , device_ctx_(device_ctx)
+        , segment_inf_(segment_inf)
+        , grid_(grid)
+        , base_cost_type_(base_cost_type)
+        , empty_(device_ctx.arch->strings.intern_string(vtr::string_view(""))) {
+        for (auto& type : physical_tile_types) {
+            tile_type_to_pb_type[std::string(type.name)] = &type;
+        }
+
+        for (size_t i = 0; i < segment_inf.size(); ++i) {
+            wire_name_to_seg_idx_[segment_inf[i].name] = i;
+        }
+
+        for (size_t i = 1; i < grid.width() - 1; ++i) {
+            for (size_t j = 1; j < grid.height() - 1; ++j) {
+                loc_ loc{i, j};
+                if (grid[i][j].type->index == 0)
+                    empty_tiles.insert(loc);
+            }
+        }
+    }
+
+    void build_rr_graph() {
+        create_switch_inf();
+        create_tile_id_to_loc();
+        create_uniq_node_ids();
+        create_sink_source_nodes();
+        create_pips_();
+        process_nodes();
+        pack_to_rr_graph();
+#ifdef DEBUG
+        print_();
+#endif
+        finish();
+    }
+
+  private:
+    Device::Reader& ar_;
+    DeviceContext& device_ctx_;
+    const std::vector<t_segment_inf>& segment_inf_;
+    const DeviceGrid& grid_;
+    const enum e_base_cost_type base_cost_type_;
+
+    vtr::interned_string empty_;
+
+    std::unordered_map<std::string, const t_physical_tile_type*> tile_type_to_pb_type;
+
+    std::unordered_map<std::string, int> wire_name_to_seg_idx_;
+
+    std::set<loc_> empty_tiles;
+
+    std::map<int /*tile_id(name)*/, loc_> tile_to_loc_;
+    std::map<loc_, int /*tile_id(name)*/> loc_to_tile_;
+
+    std::map<std::tuple<int /*timing_model*/, bool /*buffered*/>, int /*idx*/> pips_models_;
+    std::map<std::tuple<int /*node_id*/, int /*node_id*/>, std::tuple<int /*switch_id*/, std::tuple<int /*tile_id*/, int /*wire0*/, int /*wire1*/, bool /*forward*/>>> pips_;
+
+    std::map<std::tuple<int, int>, int> wire_to_node_;
+    std::map<int, std::set<loc_>> node_to_locs_;
+    std::map<std::tuple<int, int> /*<node, tile_id>*/, int /*segment type*/> node_tile_to_segment_;
+
+    /*
+     * Intermediate data storage.
+     * - Shorts represent connections between rr_nodes in a single FPGA Interchange node.
+     * - Redirect is map from node_id at location to location channel and index of track in that channel.
+     * It's useful for ends of the nodes (FPGA Interchange) that don't have representation in CHANS.
+     * - CHAN_loc_map maps from location and CHAN to vector containing tracks description.
+     */
+
+    std::vector<std::tuple<loc_, e_rr_type, int /*idx*/, loc_, e_rr_type, int /*idx*/>> shorts_;
+    std::map<std::tuple<int /*node_id*/, loc_>, std::tuple<loc_, e_rr_type /*CHANX/CHANY*/, int /*idx*/>> redirect_;
+    std::map<std::tuple<loc_, e_rr_type /*CHANX/CHANY*/>, std::vector<std::tuple<int, float, float>> /*idx = ptc,<seg_idx, R, C>*/> CHAN_loc_map_;
```

Comment:
```suggestion
    std::map<std::tuple<loc_, e_rr_type /*CHANX/CHANY*/>, std::vector<std::tuple<int, float, float>> /*idx = ptc,<seg_idx, R, C>*/> chan_loc_map_;
```

I think it is better to leave variable names all in the same style, just to reduce confusion later on where this map is used.

On an additional note, you may consider whether it is worth creating additional structs instead of having tuples, just to improve maintainability. E.g. it is easier to update a struct rather than modify a tuple definition.

Reply:
In addition, you may also introduce hash functions in a struct and be then able to use `unordered_map` instead of `map` to improve on performance.